src/StdLib.jl

```diff
@@ -36,14 +36,35 @@ Base.ncodeunits(s::CppBasicString)::Int = cppsize(s)
 Base.codeunit(s::StdString) = UInt8
 Base.codeunit(s::StdWString) = Cwchar_t == Int32 ? UInt32 : UInt16
 Base.codeunit(s::CppBasicString, i::Integer) = reinterpret(codeunit(s), cxxgetindex(s,i))
-Base.isvalid(s::CppBasicString, i::Integer) = (0 < i <= ncodeunits(s))
-function Base.iterate(s::CppBasicString, i::Integer=1)
-  if !isvalid(s,i)
-    return nothing
-  end
-  return(convert(Char,codeunit(s,i)),i+1)
+Base.isvalid(s::CppBasicString, i::Int) = checkbounds(Bool, s, i) && thisind(s, i) == i
+Base.thisind(s::CppBasicString, i::Int) = Base._thisind_str(s, i)
+Base.nextind(s::CppBasicString, i::Int) = Base._nextind_str(s, i)
+
+function Base.iterate(s::CppBasicString, i::Integer=firstindex(s))
+  i > ncodeunits(s) && return nothing
+  return convert(Char, codeunit(s, i)), nextind(s, i)
+end
+
+function Base.iterate(s::StdString, i::Integer=firstindex(s))
+  i > ncodeunits(s) && return nothing
+  j = isvalid(s, i) ? nextind(s, i) : i + 1
+  u = UInt32(codeunit(s, i)) << 24
+  (i += 1) < j || @goto ret
+  u |= UInt32(codeunit(s, i)) << 16
+  (i += 1) < j || @goto ret
+  u |= UInt32(codeunit(s, i)) << 8
+  (i += 1) < j || @goto ret
+  u |= UInt32(codeunit(s, i))
+  @label ret
+  return reinterpret(Char, u), j
+end
```

Comment:
this seems pretty cursed, is it copy-paste from somewhere?  if so add a comment with a breadcrumb, if not, provide a comment explaining what's going on 😬

Reply:
Here's the context around this. The base Julia code for string [iteration is specific to `String`](https://github.com/JuliaLang/julia/blob/01f6c4c8b86f62af923a600327f9b6be1a905193/base/strings/string.jl#L399-L428). This is rather unfortunate as any UTF-8 string type with a `UInt8` code unit I think would work in the same way. I noticed that the base `nextind` works on any string type so I used that information to convert the code units into the big endian bytes used for the UTF-8 representation taking inspiration from `Base.iterate(::String, ...)` without copying all of the baggage. Later I discovered that [WeakRefStrings.jl straight up copied the base string iteration methods](https://github.com/JuliaData/WeakRefStrings.jl/blob/c71add4c86696da63a3717df4a962e7116c1e299/src/WeakRefStrings.jl#L101-L135) but I decided to leave my version as is since it's simpler and is extensively tested here.

I'll leave a comment with the necessary context.

Reply:
Done